Firestore/core/src/firebase/firestore/model/field_value.h

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_FIELD_VALUE_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_FIELD_VALUE_H_
+
+#include <vector>
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+/**
+ * tagged-union class representing an immutable data value as stored in
+ * Firestore. FieldValue represents all the different kinds of values
+ * that can be stored in fields in a document.
+ */
+class FieldValue {
+ public:
+  /**
+   * All the different kinds of values that can be stored in fields in
+   * a document. The types of the same comparison order should be defined
+   * together as a group. The order of each group is defined by the Firestore
+   * backend.
+   */
+  enum class Type {
+    Null,  // Null
+    Boolean,  // Boolean
+    Long,  // Number type starts here
+    Double,
+    Timestamp,  // Timestamp type starts here
+    ServerTimestamp,
+    String,  // String
+    Blob,  // Blob
+    Reference,  // Reference
+    GeoPoint,  // GeoPoint
+    Array,  // Array
+    Object,  // Object
+    // New enum should not always been added at the tail. Add it to the correct
+    // position instead, see the doc comment above.
+  };
+
+  union UnionValue {
+    // There is no null type as tag_ alone is enough for Null FieldValue.
+    bool boolean_value_;
+    std::vector<FieldValue> array_value_;
+
+    // Explicitly define constructor and destructor as the default ones are
+    // deleted. In particular, for each non-POD type, there is a constructor to
+    // initialize that type.
+    UnionValue() {}
+    UnionValue(bool value) : boolean_value_(value) {}
+    UnionValue(const std::vector<FieldValue>& value) : array_value_(value) {}
+
+    // UnionValue itself does not do destruction. Instead its FieldValue
+    // container does destruction for it.
+    ~UnionValue() {}
+  };
+
+  FieldValue() : tag_(Type::Null) {}
+
+  explicit FieldValue(const std::vector<FieldValue>& value)
+      : tag_(Type::Array), value_(value) {}
+
+  FieldValue(const FieldValue& value);
+
+  ~FieldValue();
+
+  FieldValue& operator=(const FieldValue& value);
+
+  /** Returns the true type for this value. */
+  Type type() const {
+    return tag_;
+  }
+
+  /** Returns constants. */
+  static const FieldValue& NullValue();
+  static const FieldValue& BooleanValue(bool value);
```

Comment:
The javascript SDK [defines `BooleanValue::TRUE`](https://github.com/firebase/firebase-js-sdk/blob/a55dba2e140e866abe7c951551d1629cbd9d32a5/packages/firestore/src/model/field_value.ts#L160).

For consistency's sake we should probably have

```c++
static const FieldValue& True();
static const FieldValue& False();
```

in addition to this method. As in the JS SDK, this implementation can delegate to those as appropriate.